iwyu_output.cc

```diff
@@ -1426,6 +1426,71 @@ void CalculateIwyuForFullUse(OneUse* use,
+// transitively included by some include line `b` that is desired and present and
+// adds `a`'s symbol counts to `b`.
+vector<OneIncludeOrForwardDeclareLine> PruneTransitivelyPresent(
+    const vector<OneIncludeOrForwardDeclareLine>& lines,
+    const IwyuPreprocessorInfo* preprocessor_info) {
+  // In the first step, we copy all lines that are an include line, not present,
+  // desired, and whose include_file is transitively included by another line
+  // which is a desired include line into a range `indirect` and the remaining
+  // ones into another range `direct`. This is because elements of range
+  // `indirect` are then precisely those whose symbols are already exposed
+  // through a different header that is to be included anyway, so they can be
+  // omitted.
+  const auto is_transitively_present =
+      [&lines,
+       preprocessor_info](const OneIncludeOrForwardDeclareLine& target) {
+        if (!target.IsIncludeLine() || target.is_present() ||
+            !target.is_desired()) {
+          return false;
+        }
+        // Check if target is transitively included by some relevant source.
+        // This must be proper, equality does not suffice.
+        for (const OneIncludeOrForwardDeclareLine& source : lines) {
+          if (source.included_file() == target.included_file() ||
+              !source.is_desired() || !source.IsIncludeLine()) {
+            continue;
+          }
+
+          if (preprocessor_info->FileTransitivelyIncludes(
+                  source.included_file(), target.included_file())) {
+            return true;
+          }
+        }
+        return false;
+      };
+
+  vector<OneIncludeOrForwardDeclareLine> indirect, direct;
+  indirect.reserve(lines.size());
+  direct.reserve(lines.size());
+  std::partition_copy(lines.cbegin(), lines.cend(), std::back_inserter(indirect),
+                      std::back_inserter(direct), is_transitively_present);
+  // Now add the symbol uses from elements of range indirect to the
+  // corresponding source elements of range direct so that these sources are now
+  // listed as including the corresponding symbols because they do transitively.
+  for (const OneIncludeOrForwardDeclareLine& target : indirect) {
+    const auto source =
+        std::find_if(direct.begin(), direct.end(),
+                     [&target, preprocessor_info](
+                         const OneIncludeOrForwardDeclareLine& s) {
+                       return preprocessor_info->FileTransitivelyIncludes(
+                           s.included_file(), target.included_file());
+                     });
+    CHECK_(source != direct.end())
+        << "Cannot find the source for an element"
+           "that was sorted out because it had a source";
+    for (const auto& p : target.symbol_counts()) {
+      source->AddSymbolUses(p.first, p.second);
+    }
+    // source is desired now because target was.
+    CHECK_(target.is_desired()) << "Target must be desired";
```

Comment:
I'm not entirely certain what else to say?  `source` is the direct provider, `target` was indirect, and `target` was desired, so that now source is.

Reply:
Or are you talking about the `CHECK_`?  It's just an assertion for what already happened in `is_transitively_present`.

Reply:
Yes, the check message. If it ever fires, it's nice to know up-front *why* target must be desired (presumably because you only selected desired targets before?)